src: log warning on process.binding('natives')

bnoordhuis opened the pull request:
This is used by old versions of graceful-fs but will stop working in the
not too distant future.  Print a warning so people know to upgrade.

CI: https://ci.nodejs.org/job/node-test-pull-request/268/

thefourtheye commented on src/node.cc:
```diff
@@ -2234,7 +2234,21 @@ static void Binding(const FunctionCallbackInfo<Value>& args) {
+      // graceful-fs < 4 evals process.binding('natives').fs, which prohibits
+      // using internal modules in that module.  Encourage people to upgrade.
+      fprintf(stderr, "(node) process.binding('natives') is deprecated.\n");
+      fprintf(stderr, "(node) If you use graceful-fs < 4, please update.\n");
```
Should we really say this in the warning?

bnoordhuis replied:
It's the sole reason for this PR.

thefourtheye replied:
We don't know if other modules do the same. Why should we care only about graceful-fs?

bnoordhuis replied:
Because it's probably the most used offender.  The only other projects I could find with Github's (admittedly terrible) code search is an sys / util shim (which does seem to get copied around a lot, there's that.)

rvagg replied:
While I dislike calling out a specific module here, graceful-fs is really heavily used and is causing specific problems here. Perhaps the compromise is to leave the note in but put a `//TODO: Remove by v6` or similar in there just to mark it as a temporary message?

jasnell replied:
+1 @rvagg

thefourtheye commented on lib/_debugger.js:
```diff
@@ -179,7 +179,7 @@ Client.prototype._addHandle = function(desc) {
 
 
-const natives = process.binding('natives');
+const natives = process.binding('$natives');
```
`_` is the common convention for internals right?

bnoordhuis replied:
Virtual shrug.  Whether it's $ or _ doesn't seem that important to me.

brycebaril commented:
`graceful-fs` is likely the most-common but definitely not the only thing:

http://searchnod.es/#!binding(%22natives%22)/1

thefourtheye commented on src/node.cc:
```diff
@@ -2234,7 +2234,21 @@ static void Binding(const FunctionCallbackInfo<Value>& args) {
+        node::Utf8Value name(env->isolate(), frame->GetScriptName());
+        fprintf(stderr, "(node)    at %s:%d\n", *name, frame->GetLineNumber());
+      }
+      fflush(stderr);
```
stderr is not buffered right? Should we really flush it?

bnoordhuis replied:
libc I/O can be buffered, that's why the fflush() is there.

thefourtheye commented on src/node.cc:
```diff
@@ -2234,7 +2234,21 @@ static void Binding(const FunctionCallbackInfo<Value>& args) {
     DefineConstants(exports);
     cache->Set(module, exports);
-  } else if (!strcmp(*module_v, "natives")) {
+  } else if (!strcmp(*module_v, "$natives") || !strcmp(*module_v, "natives")) {
```
Isn't using strncmp better than strcmp?

brendanashworth commented:
This kind of kicks the can down the road with `process.binding`, right? Is there any feasible way we can do this with internal modules?

vkurchatkin commented:
I have an idea how we can deprecate public access to process.binding

bnoordhuis commented:
For background, the reason for this PR is that I want to pre-compile our JS code using the V8 extras infrastructure.  At that point `process.binding('natives')` can be removed if it weren't for modules like graceful-fs.

Fishrock123 commented:
I think I'm ok with this. Making guarantees about our source being exposed at run-time is a ludicrous idea.

jasnell commented:
LGTM.

bnoordhuis commented:
Thanks @jasnell.  I plan to merge this early next week so if anyone has strong feelings, now is the time to speak up.

Fishrock123 commented:
I have strong feelings about this being a good idea.

vkurchatkin commented:
@bnoordhuis I propose to use https://github.com/nodejs/node/pull/2768 for this. It's not necessary to deprecate whole `process.binding` right away, but we can use introduced hooks to deprecate `natives` in javascript. Also someone could just replace `natives` with `$natives` to suppress warning. This is not possible with https://github.com/nodejs/node/pull/2768

bnoordhuis commented:
Fine by me, as long as it gets done.  Closing.

ChALkeR commented:
@bnoordhuis It looks like it was actually never done, and that PR was closed.

Perhaps we should reopen this?

jasnell commented:
If this goes forward it should be updated to make use of the `emitWarning()` API.

bnoordhuis commented:
Rebased.  Added a second commit that takes a different approach: check if the topmost JS frame is a builtin script.

> If this goes forward it should be updated to make use of the emitWarning() API.

Why?  We don't do that for other warnings in C++ code either.

ChALkeR commented on src/node.cc:
```diff
@@ -2581,9 +2569,26 @@ static void Binding(const FunctionCallbackInfo<Value>& args) {
+      fprintf(stderr,
+              "(node:%d) process.binding('natives') is deprecated.\n", pid);
+      fprintf(stderr,
+              "(node:%d) If you use graceful-fs < 4, please update.\n", pid);
```
do recent versions of graceful-fs@3 still trigger this?

ChALkeR replied:
Another question: do we need to specifically note graceful-fs here? It's not the only package that uses `process.binding('natives')`, and that could lead people away from the actual issue.

bnoordhuis replied:
Flashback: https://github.com/nodejs/node/pull/2741#discussion_r38969876

ChALkeR replied:
@bnoordhuis yes, but see the list at https://github.com/nodejs/node/pull/2741#issuecomment-242514854.

bnoordhuis replied:
I guess I could check if the caller is graceful-fs and customize the error based on that but meh.  I have to be selective with my time in the next few weeks so I'm probably not going to expend too much effort on this PR.

jasnell replied:
If you don't have the time to work on it, I am willing to take it over.

jasnell commented:
> Why? We don't do that for other warnings in C++ code either.

So that it can be consistently handled as a runtime deprecation just like any other deprecation warning we print. As illustrated by #8270, it's pretty trivial and has the built-in benefit of working with `--throw-deprecation`.

ChALkeR commented:
I found 268 lines in 147 packages using `process.binding('natives')`: https://gist.github.com/ChALkeR/d8c9ff68869a1d86a3d702ce5c6d3a0a.

103 lines of that are `require(process.binding('natives').util ? 'util' : 'sys')` with syntaxical differences, without that — there are 165 files in 120 packages.

ChALkeR commented on src/node.cc:
```diff
@@ -2581,9 +2569,26 @@ static void Binding(const FunctionCallbackInfo<Value>& args) {
+    auto caller_script_id = GetCallerScriptId(env->isolate());
+    if (!IsInternalScriptId(env, caller_script_id)) {
+      // graceful-fs < 4 evals process.binding('natives').fs, which prohibits
+      // using internal modules in that module.  Encourage people to upgrade.
```
This is not accurate any more, they evaded that, and it now reevalutes internal modules with more dark magic, internal API usage, and relying on implementation details.

ChALkeR commented:
The warning looks like this now:
```console
$ ~/repo/node/node ./node_modules/.bin/gulp 
(node:11941) process.binding('natives') is deprecated.
(node:11941) If you use graceful-fs < 4, please update.
(node:11941)    at /home/chalker/repo/qmlweb/node_modules/natives/index.js:1
(node:11941)    at module.js:556
(node:11941)    at module.js:565
(node:11941)    at module.js:473
(node:11941)    at module.js:432
(node:11941)    at module.js:424
(node:11941)    at module.js:483
(node:11941)    at internal/module.js:20
(node:11941)    at /home/chalker/repo/qmlweb/node_modules/vinyl-fs/node_modules/graceful-fs/fs.js:1
(node:11941)    at module.js:556
(node:11941)    at module.js:565
(node:11941)    at module.js:473
[23:08:40] Using gulpfile ~/repo/qmlweb/gulpfile.js
```

Is this intentional?

bnoordhuis commented:
> Is this intentional?

Yes.  Please be specific if you think there is something unintentional in there.

ChALkeR commented:
@bnoordhuis Ok, I meant that the warning looks like too heavy to me, and most of those is cosumed by lines that do not bring much value (i.e. `at module.js:…`).

ChALkeR commented:
Btw, one more thing: 4 packages actually _assign_ to `process.binding('natives')[something]`, and they actually succeed in that. That happens inside tests only, though.

Could we stop that, like now, without waiting for a whole major release cycle? This shouldn't get misused even more.

vkurchatkin commented:
> Could we stop that, like now, without waiting for a whole major release cycle?

That would be nice, making props non-writable and non-configurable should be enough

jasnell commented:
> Could we stop that

+1 ... especially if we know we can do it without breaking non-test code.

bnoordhuis commented:
>  I meant that the warning looks like too heavy to me, and most of those is cosumed by lines that do not bring much value (i.e. at module.js:…).

I added the stack trace based on prior experience that if you don't, people are going to complain it's impossible to find out what is causing the warning.

ChALkeR commented:
> I added the stack trace based on prior experience that if you don't, people are going to complain it's impossible to find out what is causing the warning.

Makes sense. But could we strip out lines that match ` (internal/)?module.js:`? Those are pointing to internal files, if I understand things correctly.

bnoordhuis commented:
Added a commit that does that (but in a better way than pattern matching strings.)

ChALkeR commented:
@bnoordhuis Thanks!

ChALkeR commented:
@bnoordhuis Sorry, I have another warning message stylistic question. How hard would it be to print the `If you use graceful-fs < 4, please update.` line only for those cases where `graceful-fs` is present in the backtrace? That would remove possible confusion for those users that get this warning for reasons other than graceful-fs.

Btw — is this semver-patch or semver-major?

jasnell commented:
I'm -1 on this as is given that it is printed as a deprecation but does not properly follow the `--trace-deprecation` and `--throw-deprecation` command line flags. The behavior is inconsistent  when it simply needn't be.

ChALkeR commented:
@jasnell, would you change your mind if this would follow the command-line flags?

jasnell commented:
Yep... see https://github.com/nodejs/node/pull/2741#issuecomment-242512624

jasnell commented:
However, please keep in mind that the new fix for [graceful-fs](https://github.com/isaacs/node-graceful-fs/commit/331ac8db188d223a495584080504d6d55085dbcf) uses the [natives](https://www.npmjs.com/package/natives) module which uses `process.binding('natives')`, which is what has allowed us to get past the `fs` re-evaluation issue.

ChALkeR commented:
@jasnell, older graceful-fs@3 versions used `process.binding('natives')` likewise, nothing has changed on that part. Btw, `graceful-fs@4` should be fine, and most users have already upgraded to that.

jasnell commented:
Yep, just wanted to make sure it was noted.

bnoordhuis commented:
> I'm -1 on this as is given that it is printed as a deprecation but does not properly follow the --trace-deprecation and --throw-deprecation command line flags. The behavior is inconsistent when it simply needn't be.

The JS API has no way to do the internal stack frame filtering that @ChALkeR requested, and besides, other deprecation warnings from C++ (e.g. deprecated Template::Set usage) don't use it either.

I could argue the PR in its current shape is more consistent with existing practices than it would be if I included your suggestion.

jasnell commented:
At the very least `--throw-deprecation` and `--trace-deprecation` should be supported here in order for it to be a proper runtime deprecation... regardless of whether it uses `emitWarning()`.

ChALkeR commented:
@bnoordhuis 
> The JS API has no way to do the internal stack frame filtering that @ChALkeR requested

Just a note: I would be fine with not filtering if printing the trace would be behind a flag. I don't have any specific opinion on printing from js vs printing from the C++ side here at the moment, though.
